File: Tests/Tests/NSDictionary+StripeTest.m
//
//  NSDictionary+StripeTest.m
//  Stripe
//
//  Created by Joey Dong on 7/24/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "NSDictionary+Stripe.h"

@interface NSDictionary_StripeTest : XCTestCase

@end

@implementation NSDictionary_StripeTest

#pragma mark - dictionaryByRemovingNullsValidatingRequiredFields

- (void)test_dictionaryByRemovingNulls_removesNullsDeeply {
    NSDictionary *dictionary = @{
                                 @"id": @"card_123",
                                 @"tokenization_method": [NSNull null], // null in root
                                 @"metadata": @{
                                         @"user": @"user_123",
                                         @"country": [NSNull null], // null in dictionary
                                         @"nicknames": @[
                                                 @"john",
                                                 @"johnny",
                                                 [NSNull null], // null in array in dictionary
                                                 ],
                                         @"profiles": @{
                                                 @"facebook": @"fb_123",
                                                 @"twitter": [NSNull null], // null in dictionary in dictionary
                                                 }
                                         },
                                 @"fees": @[
                                         [NSNull null], // null in array
                                         @{
                                             @"id": @"fee_123",
                                             @"frequency": [NSNull null], // null in dictionary in array
                                             },
                                         @[
                                             @"payment",
                                             [NSNull null], // null in array in array
                                             ],
                                         ],
                                 };

    NSDictionary *expected = @{
                               @"id": @"card_123",
                               @"metadata": @{
                                       @"user": @"user_123",
                                       @"nicknames": @[
                                               @"john",
                                               @"johnny",
                                               ],
                                       @"profiles": @{
                                               @"facebook": @"fb_123",
                                               },
                                       },
                               @"fees": @[
                                       @{
                                           @"id": @"fee_123",
                                           },
                                       @[
                                           @"payment",
                                           ],
                                       ],
                               };

    NSDictionary *result = [dictionary stp_dictionaryByRemovingNulls];

    XCTAssertEqualObjects(result, expected);
}

- (void)test_dictionaryByRemovingNullsValidatingRequiredFields_keepsEmptyLeaves {
    NSDictionary *dictionary = @{@"id": [NSNull null]};
    NSDictionary *result = [dictionary stp_dictionaryByRemovingNulls];

    XCTAssertEqualObjects(result, @{});
}

- (void)test_dictionaryByRemovingNullsValidatingRequiredFields_returnsImmutableCopy {
    NSDictionary *dictionary = @{@"id": @"card_123"};
    NSDictionary *result = [dictionary stp_dictionaryByRemovingNulls];

    XCTAssert(result);
    XCTAssertNotEqual(result, dictionary);
    XCTAssertFalse([result isKindOfClass:[NSMutableDictionary class]]);
}

#pragma mark - dictionaryByRemovingNonStrings

- (void)test_dictionaryByRemovingNonStrings_basicCases {
    NSDictionary *dictionary;
    NSDictionary *expected;
    NSDictionary *result;

    // Empty dictionary
    dictionary = @{};
    expected = @{};
    result = [dictionary stp_dictionaryByRemovingNonStrings];
    XCTAssertEqualObjects(result, expected);

    // Regular case
    dictionary = @{
                   @"user": @"user_123",
                   @"nicknames": @"John, Johnny",
                   };
    expected = @{
                 @"user": @"user_123",
                 @"nicknames": @"John, Johnny",
                 };
    result = [dictionary stp_dictionaryByRemovingNonStrings];
    XCTAssertEqualObjects(result, expected);

    // Strips non-NSString keys and values
    dictionary = @{
                   @"user": @"user_123",
                   @"nicknames": @"John, Johnny",
                   @"profiles": [NSNull null],
                   [NSNull null]: @"San Francisco, CA",
                   [NSNull null]: [NSNull null],
                   @"age": @(21),
                   @(21): @"age",
                   @(21): @(21),
                   @"fees": @{
                           @"plan": @"monthly",
                           },
                   @"visits": @[
                           @"january",
                           @"february",
                           ],
                   };
    expected = @{
                 @"user": @"user_123",
                 @"nicknames": @"John, Johnny",
                 };
    result = [dictionary stp_dictionaryByRemovingNonStrings];
    XCTAssertEqualObjects(result, expected);
}

- (void)test_dictionaryByRemovingNonStrings_returnsImmutableCopy {
    NSDictionary *dictionary = @{@"user": @"user_123"};
    NSDictionary *result = [dictionary stp_dictionaryByRemovingNonStrings];

    XCTAssert(result);
    XCTAssertNotEqual(result, dictionary);
    XCTAssertFalse([result isKindOfClass:[NSMutableDictionary class]]);
}

#pragma mark - Getters

- (void)testArrayForKey {
    NSDictionary *dict = @{
                           @"a": @[@"foo"],
                           };

    XCTAssertEqualObjects([dict stp_arrayForKey:@"a"], @[@"foo"]);
    XCTAssertNil([dict stp_arrayForKey:@"b"]);
}

- (void)testBoolForKey {
    NSDictionary *dict = @{
                           @"a": @1,
                           @"b": @0,
                           @"c": @"true",
                           @"d": @"false",
                           @"e": @"1",
                           @"f": @"foo",
                           };

    XCTAssertTrue([dict stp_boolForKey:@"a" or:NO]);
    XCTAssertFalse([dict stp_boolForKey:@"b" or:YES]);
    XCTAssertTrue([dict stp_boolForKey:@"c" or:NO]);
    XCTAssertFalse([dict stp_boolForKey:@"d" or:YES]);
    XCTAssertTrue([dict stp_boolForKey:@"e" or:NO]);
    XCTAssertFalse([dict stp_boolForKey:@"f" or:NO]);
}

- (void)testIntForKey {
    NSDictionary *dict = @{
                           @"a": @1,
                           @"b": @-1,
                           @"c": @"1",
                           @"d": @"-1",
                           @"e": @"10.0",
                           @"f": @"10.5",
                           @"g": @(10.0),
                           @"h": @(10.5),
                           @"i": @"foo",
                           };

    XCTAssertEqual([dict stp_intForKey:@"a" or:0], 1);
    XCTAssertEqual([dict stp_intForKey:@"b" or:0], -1);
    XCTAssertEqual([dict stp_intForKey:@"c" or:0], 1);
    XCTAssertEqual([dict stp_intForKey:@"d" or:0], -1);
    XCTAssertEqual([dict stp_intForKey:@"e" or:0], 10);
    XCTAssertEqual([dict stp_intForKey:@"f" or:0], 10);
    XCTAssertEqual([dict stp_intForKey:@"g" or:0], 10);
    XCTAssertEqual([dict stp_intForKey:@"h" or:0], 10);
    XCTAssertEqual([dict stp_intForKey:@"i" or:0], 0);
}

- (void)testDateForKey {
    NSDictionary *dict = @{
                           @"a": @0,
                           @"b": @"0",
                           };
    NSDate *expectedDate = [NSDate dateWithTimeIntervalSince1970:0];

    XCTAssertEqualObjects([dict stp_dateForKey:@"a"], expectedDate);
    XCTAssertEqualObjects([dict stp_dateForKey:@"b"], expectedDate);
    XCTAssertNil([dict stp_dateForKey:@"c"]);
}

- (void)testDictionaryForKey {
    NSDictionary *dict = @{
                           @"a": @{@"foo": @"bar"},
                           };

    XCTAssertEqualObjects([dict stp_dictionaryForKey:@"a"], @{@"foo": @"bar"});
    XCTAssertNil([dict stp_dictionaryForKey:@"b"]);
}

- (void)testNumberForKey {
    NSDictionary *dict = @{
                           @"a": @1,
                           };

    XCTAssertEqualObjects([dict stp_numberForKey:@"a"], @1);
    XCTAssertNil([dict stp_numberForKey:@"b"]);
}

- (void)testStringForKey {
    NSDictionary *dict = @{@"a": @"foo"};
    XCTAssertEqualObjects([dict stp_stringForKey:@"a"], @"foo");
    XCTAssertNil([dict stp_stringForKey:@"b"]);
}

- (void)testURLForKey {
    NSDictionary *dict = @{
                           @"a": @"https://example.com",
                           @"b": @"not a url"
                           };
    XCTAssertEqualObjects([dict stp_urlForKey:@"a"], [NSURL URLWithString:@"https://example.com"]);
    XCTAssertNil([dict stp_urlForKey:@"b"]);
    XCTAssertNil([dict stp_urlForKey:@"c"]);
}

@end

File: Tests/Tests/NSDictionary+StripeTest.swift
//
//  NSDictionary+StripeTest.swift
//  Stripe
//
//  Created by Joey Dong on 7/24/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@testable import Stripe

class NSDictionary_StripeTest: XCTestCase {
  // MARK: - dictionaryByRemovingNullsValidatingRequiredFields
  func test_dictionaryByRemovingNulls_removesNullsDeeply() {
    let dictionary =
      [
        "id": "card_123",
        "tokenization_method": NSNull() /* null in root */,
        "metadata": [
          "user": "user_123",
          "country": NSNull() /* null in dictionary */,
          "nicknames": ["john", "johnny", NSNull()],
          "profiles": [
            "facebook": "fb_123",
            "twitter": NSNull(),
          ],
        ],
        "fees": [
          NSNull() /* null in array */,
          [
            "id": "fee_123",
            "frequency": NSNull(),
          ],
          ["payment", NSNull()],
        ],
      ] as NSDictionary

    let expected =
      [
        "id": "card_123",
        "metadata": [
          "user": "user_123",
          "nicknames": ["john", "johnny"],
          "profiles": [
            "facebook": "fb_123"
          ],
        ],
        "fees": [
          [
            "id": "fee_123"
          ], ["payment"],
        ],
      ] as NSDictionary

    let result = dictionary.stp_dictionaryByRemovingNulls() as NSDictionary
    XCTAssertEqual(result, expected)
  }

  func test_dictionaryByRemovingNullsValidatingRequiredFields_keepsEmptyLeaves() {
    let dictionary =
      [
        "id": NSNull()
      ] as NSDictionary
    let result = dictionary.stp_dictionaryByRemovingNulls() as NSDictionary

    XCTAssertEqual(result, [:])
  }

  // MARK: - dictionaryByRemovingNonStrings
  func test_dictionaryByRemovingNonStrings_basicCases() {
    // Empty dictionary
    var dictionary = [:] as NSDictionary
    var expected = [:] as NSDictionary
    var result = dictionary.stp_dictionaryByRemovingNonStrings() as NSDictionary
    XCTAssertEqual(result, expected)

    // Regular case
    dictionary =
      [
        "user": "user_123",
        "nicknames": "John, Johnny",
      ] as NSDictionary
    expected =
      [
        "user": "user_123",
        "nicknames": "John, Johnny",
      ] as NSDictionary
    result = dictionary.stp_dictionaryByRemovingNonStrings() as NSDictionary
    XCTAssertEqual(result, expected)

    // Strips non-NSString keys and values
    dictionary =
      [
        "user": "user_123",
        "nicknames": "John, Johnny",
        "profiles": NSNull(),
        NSNull(): "San Francisco, CA",
        NSNull(): NSNull(),
        "age": NSNumber(value: 21),
        NSNumber(value: 21): "age",
        NSNumber(value: 21): NSNumber(value: 21),
        "fees": [
          "plan": "monthly"
        ],
        "visits": ["january", "february"],
      ] as NSDictionary
    expected =
      [
        "user": "user_123",
        "nicknames": "John, Johnny",
      ] as NSDictionary
    result = dictionary.stp_dictionaryByRemovingNonStrings() as NSDictionary
    XCTAssertEqual(result, expected)
  }

  // MARK: - Getters
  func testArrayForKey() {
    let dict =
      [
        "a": ["foo"]
      ] as NSDictionary

    XCTAssertEqual(dict.stp_array(forKey: "a") as! [String], ["foo"])
    XCTAssertNil(dict.stp_array(forKey: "b"))
  }

  func testBoolForKey() {
    let dict =
      [
        "a": NSNumber(value: 1),
        "b": NSNumber(value: 0),
        "c": "true",
        "d": "false",
        "e": "1",
        "f": "foo",
      ] as NSDictionary

    XCTAssertTrue(dict.stp_bool(forKey: "a", or: false))
    XCTAssertFalse(dict.stp_bool(forKey: "b", or: true))
    XCTAssertTrue(dict.stp_bool(forKey: "c", or: false))
    XCTAssertFalse(dict.stp_bool(forKey: "d", or: true))
    XCTAssertTrue(dict.stp_bool(forKey: "e", or: false))
    XCTAssertFalse(dict.stp_bool(forKey: "f", or: false))
  }

  func testIntForKey() {
    let dict =
      [
        "a": NSNumber(value: 1),
        "b": NSNumber(value: -1),
        "c": "1",
        "d": "-1",
        "e": "10.0",
        "f": "10.5",
        "g": NSNumber(value: 10.0),
        "h": NSNumber(value: 10.5),
        "i": "foo",
      ] as NSDictionary

    XCTAssertEqual(dict.stp_int(forKey: "a", or: 0), 1)
    XCTAssertEqual(dict.stp_int(forKey: "b", or: 0), -1)
    XCTAssertEqual(dict.stp_int(forKey: "c", or: 0), 1)
    XCTAssertEqual(dict.stp_int(forKey: "d", or: 0), -1)
    XCTAssertEqual(dict.stp_int(forKey: "e", or: 0), 10)
    XCTAssertEqual(dict.stp_int(forKey: "f", or: 0), 10)
    XCTAssertEqual(dict.stp_int(forKey: "g", or: 0), 10)
    XCTAssertEqual(dict.stp_int(forKey: "h", or: 0), 10)
    XCTAssertEqual(dict.stp_int(forKey: "i", or: 0), 0)
  }

  func testDateForKey() {
    let dict =
      [
        "a": NSNumber(value: 0),
        "b": "0",
      ] as NSDictionary
    let expectedDate = Date(timeIntervalSince1970: 0)

    XCTAssertEqual(dict.stp_date(forKey: "a"), expectedDate)
    XCTAssertEqual(dict.stp_date(forKey: "b"), expectedDate)
    XCTAssertNil(dict.stp_date(forKey: "c"))
  }

  func testDictionaryForKey() {
    let dict =
      [
        "a": [
          "foo": "bar"
        ]
      ] as NSDictionary

    XCTAssertEqual(
      dict.stp_dictionary(forKey: "a")! as NSDictionary,
      [
        "foo": "bar"
      ])
    XCTAssertNil(dict.stp_dictionary(forKey: "b"))
  }

  func testNumberForKey() {
    let dict =
      [
        "a": NSNumber(value: 1)
      ] as NSDictionary

    XCTAssertEqual(dict.stp_number(forKey: "a"), NSNumber(value: 1))
    XCTAssertNil(dict.stp_number(forKey: "b"))
  }

  func testStringForKey() {
    let dict =
      [
        "a": "foo"
      ] as NSDictionary
    XCTAssertEqual(dict.stp_string(forKey: "a"), "foo")
    XCTAssertNil(dict.stp_string(forKey: "b"))
  }

  func testURLForKey() {
    let dict =
      [
        "a": "https://example.com",
        "b": "not a url",
      ] as NSDictionary
    XCTAssertEqual(dict.stp_url(forKey: "a"), URL(string: "https://example.com"))
    XCTAssertNil(dict.stp_url(forKey: "b"))
    XCTAssertNil(dict.stp_url(forKey: "c"))
  }
}
